recipe-server/client/control/components/MultiPicker.js

```diff
@@ -51,88 +44,65 @@ export default class MultiPicker extends React.Component {
       value,
     } = this.props;
 
-    const selectedOptions = options.filter(option =>
-      value.indexOf(option.value) === -1);
-
-    let displayedOptions = [].concat(selectedOptions)
-      .map(option => option.value)
-      .map(val =>
-        (!val ? null : { ...this.convertValueToObj(val) }))
-      .filter(x => x);
-
     const {
       filterText,
     } = this.state;
 
-    if (filterText) {
-      displayedOptions = displayedOptions.filter(option =>
-        JSON.stringify(option).indexOf(filterText) > -1);
-    }
+    let selectedOptions = options.filter(option => value.indexOf(option.value) === -1);
 
-    return displayedOptions;
-  }
-
-  handleApplyOption(event) {
-    event.persist();
-
-    if (!this.availableRef) {
-      return;
-    }
-
-    const {
-      value,
-      onChange,
-    } = this.props;
-
-    let selectedFilters = MultiPicker.getActiveSelectOptions(this.availableRef);
+    if (filterText) {
+      const searchText = filterText.toLowerCase();
 
-    if (!selectedFilters || selectedFilters.length === 0) {
-      selectedFilters = this.getDisplayedOptions();
-      selectedFilters = selectedFilters.map(option => option.value);
+      selectedOptions = selectedOptions.filter(option => {
+        const searchValues = JSON.stringify(Object.values(option)).toLowerCase();
+        return searchValues.indexOf(searchText) > -1;
+      });
     }
 
-    const newOptions = []
-      .concat(value || [])
-      .concat(selectedFilters);
-
-    // clear user input
-    this.availableRef.value = null;
-
-    onChange(newOptions.join(','));
+    return selectedOptions;
   }
 
-  handleRemoveOption(event) {
-    event.persist();
-    if (!this.selectedRef) {
-      return;
-    }
-
-    const {
-      value,
-      onChange,
-    } = this.props;
+  handleTextChange(event) {
+    const { value } = event.target;
 
-    const selectedFilters = MultiPicker.getActiveSelectOptions(this.selectedRef);
-    let newOptions = []
-      .concat(value || [])
-      .filter(val => selectedFilters.indexOf(val) === -1);
+    this.setState({
+      filterText: value,
+    });
+  }
 
-    if (!selectedFilters || selectedFilters.length === 0) {
-      newOptions = [];
-    }
+  handleChangeOption(type) {
+    return () => {
+      const ref = type === 'apply' ? this.availableRef : this.selectedRef;
+      if (!type || !ref) {
+        return;
+      }
 
-    // clear the user selection
-    this.selectedRef.value = null;
+      const {
+        value = [],
+        onChange = () => {},
+      } = this.props;
+
+      // Get a list of all selected values.
+      let selection = [].concat(value);
+      // Pick up list of selected options from the dom ref
+      const refSelected = MultiPicker.getActiveSelectOptions(ref);
+
+      // Apply = take from the 'available' section and put in 'selected'
+      if (type === 'apply') {
```

Comment:
Sorry, can you elaborate on `Deferring to another method on the class and subclassing to customize the behavior.` ? I'm unsure how this function could be split up, as both operations (applying/removing) need to perform mostly the same operations, in the same order. Any clarity would be great, thank you!